pallets/xcmp-handler/src/lib.rs

```diff
@@ -0,0 +1,160 @@
+			+ MaybeSerializeDeserialize
+			+ Ord
+			+ TypeInfo
+			+ MaxEncodedLen;
+
+		/// The currencyId for the native currency.
+		#[pallet::constant]
+		type GetNativeCurrencyId: Get<Self::CurrencyId>;
+
+		/// Weight information for extrinsics in this module.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// XCM data was added for a chain/currency pair.
+		XcmDataAdded { para_id: ParachainId, currency_id: T::CurrencyId },
+		/// XCM data was removed for a chain/currency pair.
+		XcmDataRemoved { para_id: ParachainId, currency_id: T::CurrencyId },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// We only support certain currency/chain combinations.
+		CurrencyChainComboNotSupported,
+		/// There is no entry for that currency/chain combination.
+		CurrencyChainComboNotFound,
+	}
+
+	/// Stores all data needed to send an XCM message for chain/currency pair.
+	#[derive(Clone, Debug, Encode, Decode, PartialEq, TypeInfo)]
+	pub struct XcmCurrencyData {
+		/// Is the token native to the chain?
+		pub native: bool,
+		pub fee_per_second: u128,
+		/// The weight of the instructions for the chain/currency pair minus the Transact.
+		pub instruction_weight: u64,
+	}
+
+	/// Stores XCM data for a chain/currency pair.
+	#[pallet::storage]
+	#[pallet::getter(fn get_xcm_chain_data)]
+	pub type XcmChainCurrencyData<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		ParachainId,
+		Twox64Concat,
+		T::CurrencyId,
+		XcmCurrencyData,
+	>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Add or update XCM data for a chain/currency pair.
+		/// For now we only support our native currency.
+		#[pallet::weight(T::WeightInfo::add_chain_currency_data())]
+		pub fn add_chain_currency_data(
+			origin: OriginFor<T>,
+			para_id: ParachainId,
+			currency_id: T::CurrencyId,
+			xcm_data: XcmCurrencyData,
+		) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+
+			if currency_id != T::GetNativeCurrencyId::get() {
+				Err(Error::<T>::CurrencyChainComboNotSupported)?
+			}
+
+			XcmChainCurrencyData::<T>::insert(para_id, currency_id, xcm_data);
+			Self::deposit_event(Event::XcmDataAdded { para_id, currency_id });
+
+			Ok(().into())
+		}
+
+		/// Remove XCM data for a chain/currency pair.
+		#[pallet::weight(T::WeightInfo::remove_chain_currency_data())]
+		pub fn remove_chain_currency_data(
+			origin: OriginFor<T>,
+			para_id: ParachainId,
+			currency_id: T::CurrencyId,
+		) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+
+			if let Some(_) = XcmChainCurrencyData::<T>::get(para_id, currency_id) {
+				XcmChainCurrencyData::<T>::remove(para_id, currency_id);
+				Self::deposit_event(Event::XcmDataRemoved { para_id, currency_id });
+			} else {
+				Err(Error::<T>::CurrencyChainComboNotFound)?
+			}
```

Comment:
Not sure if syntactic sugar or magically saves a read. Probably just syntactic sugar.
```suggestion
			XcmChainCurrencyData::<T>::take(para_id, currency_id)
				.ok_or(Error::<T>::CurrencyChainComboNotFound)?;
			Self::deposit_event(Event::XcmDataRemoved { para_id, currency_id });
```